core/src/main/java/org/elasticsearch/bootstrap/Spawner.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.bootstrap;
+
+import org.apache.lucene.util.Constants;
+import org.elasticsearch.env.Environment;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Spawns native plugin controller processes if present.  Will only work prior to seccomp being set up.
+ */
+final class Spawner implements Closeable {
+
+    private static final String PROGRAM_NAME = Constants.WINDOWS ? "controller.exe" : "controller";
+    private static final String PLATFORM_NAME = makePlatformName(Constants.OS_NAME, Constants.OS_ARCH);
+    private static final String TMP_ENVVAR = "TMPDIR";
+
+    /**
+     * On Windows we have to retain a reference to the OutputStream that corresponds to each
+     * process's stdin, otherwise the pipe will be closed and the spawned process will receive
+     * an EOF on its stdin.  This isn't necessary on *nix for daemons that only need to shut
+     * down when the JVM exits, but storing these references improves testability by making this
+     * class Closeable.
+     */
+    private final List<OutputStream> stdinReferences = new ArrayList<>();
+
+    @Override
+    public void close() throws IOException {
+        for (OutputStream stream : stdinReferences) {
+            stream.close();
+        }
+    }
+
+    /**
+     * For each plugin, attempt to spawn the controller daemon.  Silently ignore any plugins
+     * that don't include a version of the program for the correct platform.
+     */
+    void spawnNativePluginControllers(Environment environment) throws IOException {
+        if (Files.exists(environment.pluginsFile())) {
+            try (DirectoryStream<Path> stream = Files.newDirectoryStream(environment.pluginsFile())) {
+                for (Path plugin : stream) {
+                    Path spawnPath = makeSpawnPath(plugin);
+                    if (Files.isRegularFile(spawnPath)) {
+                        spawnNativePluginController(environment, spawnPath);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Attempt to spawn the controller daemon for a given plugin.  The spawned process
+     * will remain connected to this JVM via its stdin, stdout and stderr, but the
+     * references to these streams are not available to code outside this class.
+     */
+    private void spawnNativePluginController(Environment environment, Path spawnPath) throws IOException {
```

Comment:
or rather pass it to `spawnNativePluginControllers` and change it's signature to `spawnNativePluginControllers(Path pluginPath, Map<String,String> env)` I don't think we need to depend on `Environment`